// ng-bootstrap-form-validation/esm5/lib/ng-bootstrap-form-validation.module.js
import { __decorate } from "tslib";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { FormValidationDirective } from "./Directives/form-validation.directive";
import { MessagesComponent } from "./Components/messages/messages.component";
import { ErrorMessageService } from "./Services/error-message.service";
import { CUSTOM_ERROR_MESSAGES, BOOTSTRAP_VERSION } from "./Tokens/tokens";
import { BootstrapVersion } from "./Enums/BootstrapVersion";
import { FormGroupComponent } from "./Components/form-group/form-group.component";
import { FormControlDirective } from "./Directives/form-control.directive";
var NgBootstrapFormValidationModule = /** @class */ (function () {
    function NgBootstrapFormValidationModule() {
    }
    NgBootstrapFormValidationModule_1 = NgBootstrapFormValidationModule;
    NgBootstrapFormValidationModule.forRoot = function (userOptions) {
        if (userOptions === void 0) { userOptions = {
            bootstrapVersion: BootstrapVersion.Four
        }; }
        return {
            ngModule: NgBootstrapFormValidationModule_1,
            providers: [
                {
                    provide: CUSTOM_ERROR_MESSAGES,
                    useValue: userOptions.customErrorMessages || [],
                    multi: true
                },
                {
                    provide: BOOTSTRAP_VERSION,
                    useValue: userOptions.bootstrapVersion
                },
                ErrorMessageService
            ]
        };
    };
    var NgBootstrapFormValidationModule_1;
    NgBootstrapFormValidationModule = NgBootstrapFormValidationModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [
                FormValidationDirective,
                FormGroupComponent,
                MessagesComponent,
                FormControlDirective
            ],
            exports: [
                FormValidationDirective,
                FormGroupComponent,
                MessagesComponent,
                FormControlDirective
            ]
        })
    ], NgBootstrapFormValidationModule);
    return NgBootstrapFormValidationModule;
}());
export { NgBootstrapFormValidationModule };
