// ng-bootstrap-form-validation/esm2015/lib/default-errors.js
const ɵ0 = label => `${label} is required`, ɵ1 = label => `${label} is invalid`, ɵ2 = (label, error) => `${label} must be at least ${error.requiredLength} characters`, ɵ3 = (label, error) => `${label} must be no longer than ${error.requiredLength} characters`, ɵ4 = (label, error) => `${label} is required`, ɵ5 = (label, error) => `Invalid email address`, ɵ6 = (label, error) => `${label} must be no greater than ${error.max}`, ɵ7 = (label, error) => `${label} must be no less than ${error.min}`;
export const DEFAULT_ERRORS = [
    {
        error: "required",
        format: ɵ0
    },
    {
        error: "pattern",
        format: ɵ1
    },
    {
        error: "minlength",
        format: ɵ2
    },
    {
        error: "maxlength",
        format: ɵ3
    },
    {
        error: "requiredTrue",
        format: ɵ4
    },
    {
        error: "email",
        format: ɵ5
    },
    {
        error: "max",
        format: ɵ6
    },
    {
        error: "min",
        format: ɵ7
    }
];
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7 };
